lib/ruby_lsp/requests/request_with_environment.rb

```diff
@@ -0,0 +1,64 @@
+# typed: strict
+# frozen_string_literal: true
+
+module RubyLsp
+  module Requests
+    # When creating requests, this class should be used instead of BaseRequest if the request needs information about
+    # the environment (e.g.: which locals are parameters, or all occurrences of the same local variable).
+    class RequestWithEnvironment < BaseRequest
+      extend T::Sig
+      extend T::Helpers
+
+      abstract!
+
+      sig { returns(Support::Environment) }
+      attr_reader :current_environment
+
+      sig { params(document: Document).void }
+      def initialize(document)
+        @current_environment = T.let(Support::Environment.new, Support::Environment)
+        super
+      end
+
+      sig { override.overridable.returns(Object) }
+      def run; end
+
+      sig { params(node: T.nilable(SyntaxTree::Node)).void }
+      def visit(node)
+        case node
+        when SyntaxTree::ClassDeclaration, SyntaxTree::ModuleDeclaration, SyntaxTree::MethodAddBlock, SyntaxTree::Def,
+             SyntaxTree::Defs, SyntaxTree::DefEndless
+          #  Create new environment to keep track of locals
+          with_new_environment { super }
+        when SyntaxTree::Params
+          declare_parameters(node)
+          super
+        when SyntaxTree::RestParam, SyntaxTree::KwRestParam, SyntaxTree::BlockArg
+          name = node.name
+          @current_environment.declare_argument(name) if name
+          super
+        else
+          super
+        end
+      end
+
+      private
+
+      sig { params(node: SyntaxTree::Params).void }
+      def declare_parameters(node)
+        node.requireds.each { |param| @current_environment.declare_argument(param) }
+        node.posts.each { |param| @current_environment.declare_argument(param) }
+        node.keywords.each { |param| @current_environment.declare_argument(param.first) }
+        node.optionals.each { |param| @current_environment.declare_argument(param.first) }
+      end
+
+      sig { params(block: T.proc.void).void }
+      def with_new_environment(&block)
+        previous_environment = @current_environment
+        @current_environment = Support::Environment.new(@current_environment)
+        block.call
+        @current_environment = previous_environment
```

Comment:
Do we need to put this in an `ensure` block to make sure environment is restored? I know in the case of exception that probably won't matter though.